metric/instrument/asyncfloat64/asyncfloat64.go

```diff
@@ -21,18 +21,33 @@ import (
 	"go.opentelemetry.io/otel/metric/instrument"
 )
 
+// Measurement is an recorded event value for a particular set of attributes.
+type Measurement struct {
+	Attributes []attribute.KeyValue
+	Value      float64
+}
+
+// Callback is a function that returns observations for an Asynchronous
+// instrument.
+//
+// The function needs to complete in a finite amount of time and the deadline
+// of the passed context is expected to be honored.
+//
+// The function needs to be concurrent safe.
+type Callback func(context.Context) ([]Measurement, error)
```

Comment:
[`produce`](https://github.com/open-telemetry/opentelemetry-go/blob/4e763472eee5ef03b97274694d2d1b07f0b3cd11/sdk/metric/pipeline.go#L114) already returns an error to the operator. This error can be propagated there. By doing this the operator of code instrumented with OpenTelemetry can make the decision on how to handle errors instead of instrumentation authors who will define the callback. By having instrumentation authors decide how errors are handled operators are not able to change that decision, especially if they want to retry the collect for manual readers.

Reply:
So are you saying that if there is one or more Callback that has an error we should still try to build the total `ResourceMetrics`? 

Do we still want to have a special case if the context is ended?
  If we do, how do we determine and broadcast our context ended, and have it be separate from a callback's internal context ended?
  For Exporter authors when should they not attempt an export and when should they attempt one?
Do we want library authors to be able to put inconsequential errors (meaning there were problems, but there still is data) in the path of a different author's Exporter/Reader?
If there are some errors that should be logged only (don't stop exports) should it be the responsibility of the exporters to do the logging?

I don't think we should just bundle these errors with the one error that `produce()` creates.  Because we know exactly what that error should do: because the context is closed no more processing.  When we allow any error we now have to handle any error, and I think in all the external cases we don't need to take any different action.

Reply:
> So are you saying that if there is one or more Callback that has an error we should still try to build the total `ResourceMetrics`?

Yes, I think a partial success here would be more useful to users than a complete drop of data.

Reply:
> * Do we still want to have a special case if the context is ended?
>   * If we do, how do we determine and broadcast our context ended, and have it be separate from a callback's internal context ended?

I don't see why not. That is the point of  `error.Is` or `error.As`.

>   * For Exporter authors when should they not attempt an export and when should they attempt one?

I think that is up to the exporter. Providing them this information gives them the opportunity to decide. Not returning an error has the instrumentation and SDK make the decision for them.

Reply:
> * Do we want library authors to be able to put inconsequential errors (meaning there were problems, but there still is data) in the path of a different author's Exporter/Reader?

I don't think we want that, but it is not relevant to the discussion of the SDK propagating errors. If the instrumentation authors provide poor errors as a callback response, log them, or push them to the error handler they are all still issues between the instrumentation author and operator of the code.

Reply:
> * If there are some errors that should be logged only (don't stop exports) should it be the responsibility of the exporters to do the logging?

Again, that is a issue an exporter will need to decide. I imagine some exporters will do this, and others might try to remediate issues and then immediately retry the collection.

Reply:
> I don't think we should just bundle these errors with the one error that `produce()` creates. Because we know exactly what that error should do: because the context is closed no more processing. When we allow any error we now have to handle any error, and I think in all the external cases we don't need to take any different action.

Again, I think this is the point of `error.Is` and `error.As`.

Reply:
I would prefer that callback can return an error.  It leaves the door open for better handling in the future.  I also like that it results in all callback errors being funneled to the global error handler.  I would worry that instrumentation might choose to do something worse than using the error handler (e.g. use `fmt.Println` directly).